src/callsite.jl

```diff
@@ -517,4 +518,19 @@ end
 maybe_callsite(info::PureCallInfo, mi::MethodInstance) = mi.specTypes <: Tuple{mapany(Core.Typeof ∘ unwrapconst, info.argtypes)...}
 maybe_callsite(info::RTCallInfo, mi::MethodInstance) = false
 
+function maybe_callsite(info::RTCallInfo, @nospecialize(tt::Type))
+    isa(tt, Union) && return maybe_callsite(info, tt.a) || maybe_callsite(info, tt.b)
+    isa(tt, DataType) || return false
+    typeof(info.f) === tt.parameters[1] || return false
+    for (a, b) in zip(info.argtyps, tt.parameters[2:end])
+        a === b || return false
```

Comment:
When iterating over a `zip` created from a `Vector{Any}`, objects of type `Tuple{<:Any,<:Any}` are created, which can sometimes lead to unnecessary specializations, especially when the vector contains `Type` objects.
Since we're only using `===`, it shouldn't be a problem for this case, but it's something to be mindful of when working on `Core.Compiler` and other external abstract interpreters in the future.

Reply:
Yeah, I figured this case would be OK. I was leaning on the "self check" tests too, I admit!